File: share/web_surfaces/builtin/mixer/toolkit/modules/range.js
/*
 * This file is part of Toolkit.
 *
 * Toolkit is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Toolkit is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General
 * Public License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301  USA
 */
"use strict";
(function(w, TK) {
TK.Range = TK.class({
    /**
    * TK.Range is used for calculating linear scales from
    * different values. They are useful to build coordinate systems,
    * calculate pixel positions for different scale types and the like.
    * TK.Range is used e.g. in {@link TK.Scale}, {@link TK.MeterBase} and {@link TK.Graph} to draw
    * elements on a certain position according to a value on an
    * arbitrary scale.
    * 
    * @class TK.Range
    * 
    * @extends TK.Base
    * 
    * @mixes TK.Ranged
    * 
    * @param {Object} [options={ }] - An object containing initial options.
    * 
    * @property {String|Function} [options.scale="linear"] - Type of the value.
    *   <code>linear</code>, <code>decibel</code>, <code>log2</code>, <code>frequency</code>
    *   or a <code>function (value, options, coef) {}</code>.
    *   If a function instead of a constant is handed over, it receives the
    *   actual options object as the second argument and is supposed to return a
    *   coefficient between 0 and 1. If the third argument "coef" is true, it is
    *   supposed to return a value depending on a coefficient handed over as the 
    *   first argument.
    * @property {Boolean} [options.reverse=false] - <code>true</code> if
    *   the range is reversed.
    * @property {Number} [options.basis=0] - Dimensions of the range, set to
    *   width/height in pixels, if you need it for drawing purposes, to 100 if
    *   you need percentual values or to 1 if you just need a linear
    *   coefficient for a e.g. logarithmic scale.
    * @property {Number} [options.min=0] - The minimum value possible.
    * @property {Number} [options.max=0] - The maximum value possible.
    * @property {Number} [options.step=1] - Step size, needed for e.g. user
    *   interaction
    * @property {Number} [options.shift_up=4] - Multiplier for e.g. SHIFT pressed
    *   while stepping
    * @property {Number} [options.shift_down=0.25] - Multiplier for e.g. SHIFT + CONTROL
    *   pressed while stepping
    * @property {Number|Array} [options.snap=0] - Snap the value to a virtual grid
    *   with this distance. Numbers define the step size between snaps, an
    *   array contains a list of values to snap to.
    *   Using snap option with float values
    *   causes the range to reduce its minimum and maximum values depending
    *   on the amount of decimal digits because of the implementation of
    *   math in JavaScript. Using a step size of e.g. 1.125
    *   reduces the maximum usable value from 9,007,199,254,740,992 to
    *   9,007,199,254,740.992 (note the decimal point).
    * @property {Boolean} [options.round=false] - if snap is set,
    *   decide how to jump between snaps. Setting this to true
    *   slips to the next snap if the value is more than on its half way to it.
    *   Otherwise the value has to reach the next snap until it is hold there
    *   again.
     */
    Extends : TK.Base,
    _class: "Range",
    Implements: [TK.Ranged],
    _options: {
        scale: "string|function",
        reverse: "boolean",
        basis: "number",
        min: "number",
        max: "number",
        step: "number",
        shift_up: "number",
        shift_down: "number",
        snap: "number|array",
        round: "boolean",
    },
    options: {
        scale:      "linear",
        reverse:    false,
        basis:      0,
        min:        0,
        max:        0,
        step:       1,
        shift_up:   4, 
        shift_down: 0.25, 
        snap:       0, 
        round:      false 
    },
});
})(this, this.TK);

File: libs/temporal/range.cc
/*
    Copyright (C) 2017 Paul Davis

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "temporal/range.h"

using namespace Temporal;

/** Subtract the ranges in `sub' from this range returning the result.
 */
RangeList
Range::subtract (RangeList & sub) const
{
	/* Start with the input range */

	RangeList result;

	result.add (*this);

	if (sub.empty () || empty()) {
		return result;
	}

	RangeList::List s = sub.get ();

	/* The basic idea here is to keep a list of the result ranges, and subtract
	   the bits of `sub' from them one by one.
	*/

	for (typename RangeList::List::const_iterator i = s.begin(); i != s.end(); ++i) {

		/* Here's where we'll put the new current result after subtracting *i from it */
		RangeList new_result;

		typename RangeList::List r = result.get ();

		/* Work on all parts of the current result using this range *i */
		for (typename RangeList::List::const_iterator j = r.begin(); j != r.end(); ++j) {

			switch (coverage_exclusive_ends (j->start(), j->end(), i->start(), i->end())) {
			case OverlapNone:
				/* The thing we're subtracting (*i) does not overlap this bit of the result (*j),
				   so pass it through.
				*/
				new_result.add (*j);
				break;
			case OverlapInternal:
				/* Internal overlap of the thing we're subtracting (*i) from this bit of the result,
				   so we should end up with two bits of (*j) left over, from the start of (*j) to
				   the start of (*i), and from the end of (*i) to the end of (*j).
				*/
				assert (j->start() < i->start());
				assert (j->end() > i->end());
				new_result.add (Range (j->start(), i->start()));
				new_result.add (Range (i->end(), j->end()));
				break;
			case OverlapStart:
				/* The bit we're subtracting (*i) overlaps the start of the bit of the result (*j),
				 * so we keep only the part of of (*j) from after the end of (*i)
				 */
				assert (i->end() < j->end());
				new_result.add (Range (i->end(), j->end()));
				break;
			case OverlapEnd:
				/* The bit we're subtracting (*i) overlaps the end of the bit of the result (*j),
				 * so we keep only the part of of (*j) from before the start of (*i)
				 */
				assert (j->start() < i->start());
				new_result.add (Range (j->start(), i->start()));
				break;
			case OverlapExternal:
				/* total overlap of the bit we're subtracting with the result bit, so the
				   result bit is completely removed; do nothing */
				break;
			}
		}

		new_result.coalesce ();
		result = new_result;
	}

	return result;
}
